src/dowel/csv_output.py

```diff
@@ -48,7 +50,35 @@ def record(self, data, prefix=''):
                            'Did you change key sets after your first '
                            'logger.log(TabularInput)?'.format(
                                set(self._fieldnames), set(to_csv.keys())))
-
+                
+                # close any existing log file
+                super().close()
+
+                # Read data from the existing log file once
+                with open(self._filename, 'r') as existing_file:
+                    reader = csv.DictReader(existing_file)
+                    file_data = []
+                    for row in reader:
+                        file_data.append(row)
+
+                    # Update header with new keys by union 
+                    self._joint_filednames = set(self._fieldnames)| set(to_csv.keys())
+
+                    # Update the writer with joint headers
+                    self._writer = csv.DictWriter(
+                        self._log_file,
+                        fieldnames=self._joint_filednames,
+                        restval='',
+                        extrasaction='ignore')
```

Comment:
Do you think it would be safer to use the default value here (`extrasaction='raise'`)?

Reply:
Right, the optional 'raise' will give  a ValueError when there are extra values in the dictionary instead of ignoring it. Changed to `extrasaction='raise'`